File: instrumentation/activity/src/main/java/io/opentelemetry/android/instrumentation/activity/startup/AnchoredClock.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.activity.startup;

import io.opentelemetry.sdk.common.Clock;

// copied from otel-java
public final class AnchoredClock {
    private final Clock clock;
    private final long epochNanos;
    private final long nanoTime;

    private AnchoredClock(Clock clock, long epochNanos, long nanoTime) {
        this.clock = clock;
        this.epochNanos = epochNanos;
        this.nanoTime = nanoTime;
    }

    public static AnchoredClock create(Clock clock) {
        return new AnchoredClock(clock, clock.now(), clock.nanoTime());
    }

    public long now() {
        long deltaNanos = this.clock.nanoTime() - this.nanoTime;
        return this.epochNanos + deltaNanos;
    }
}

File: instrumentation/activity/src/main/java/io/opentelemetry/android/instrumentation/activity/startup/AnchoredClock.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.activity.startup

import io.opentelemetry.sdk.common.Clock

internal class AnchoredClock(
    private val clock: Clock,
    private val epochNanos: Long = clock.now(),
    private val nanoTime: Long = clock.nanoTime(),
) {
    fun now(): Long {
        val deltaNanos = clock.nanoTime() - nanoTime
        return epochNanos + deltaNanos
    }

    /**
     * Converts a timestamp taken from the same [System.nanoTime] timebase as this clock's anchor
     * (for example, a [android.view.Choreographer.FrameCallback] frameTimeNanos value) into an
     * epoch-nanos timestamp comparable to [now].
     */
    fun toEpochNanos(nanoTimeValue: Long): Long {
        val deltaNanos = nanoTimeValue - nanoTime
        return epochNanos + deltaNanos
    }
}
